password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+{-|
+Module      : Data.Password.Valid
+Copyright   : (c) Dennis Gosnell, 2019; Felix Paulusma, 2020; Hiroto Shioi, 2020
+License     : BSD-style (see LICENSE file)
+Maintainer  : cdep.illabout@gmail.com
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Predicate
+    CharSetPredicate(..),
+    -- * Default
+    defaultPasswordPolicy,
+    defaultCharSetPredicate,
+    -- * Functions
+    isValidPassword,
+    validatePassword,
+    -- * Utility
+    isValidPasswordPolicy,
+    isSpecial,
+    defaultCharSet
+  ) where
+
+import Data.Char (chr, isAsciiLower, isAsciiUpper, isDigit, ord)
+import Data.Maybe (catMaybes, fromMaybe, isJust)
+import Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
```

Comment:
Reminder to add a note to explain that the `maximumLength` field should be:
* `maximumLength > 0`
* `maximumLength > minimumLength`

Also, that if any other field is set to a negative number, it will be defaulted to 0.
And also advise to use the `defaultPasswordPolicy` if the user is unsure.

(I think we can also drop the `Maybe` from the `...Chars` fields, since we're handling negative numbers and `Nothing` as 0)